packages/@uppy/companion/src/server/middlewares.js

```diff
@@ -73,15 +75,41 @@ exports.loadSearchProviderToken = (req, res, next) => {
   next()
 }
 
-exports.mergeAccessControlAllowMethods = (req, res, next) => {
-  const existingHeader = res.get('Access-Control-Allow-Methods')
-  let existingMethods = []
-  if (existingHeader) {
-    existingMethods = existingHeader.replace(/\s/g, '').split(',').map((method) => method.toUpperCase())
+exports.cors = (options = {}) => (req, res, next) => {
+  const exposedHeaders = [
+    // exposed so it can be accessed for our custom uppy client preflight
+    'Access-Control-Allow-Headers',
+  ]
+  if (options.sendSelfEndpoint) exposedHeaders.push('i-am')
+  if (res.get('Access-Control-Expose-Headers')) exposedHeaders.push(res.get('Access-Control-Expose-Headers'))
+
+  const allowedHeaders = [
+    'uppy-auth-token',
+    'uppy-versions',
+    'uppy-credentials-params',
+  ]
+  if (res.get('Access-Control-Allow-Headers')) allowedHeaders.push(res.get('Access-Control-Allow-Headers'))
+
+  const existingAllowMethodsHeader = res.get('Access-Control-Allow-Methods')
+  let methods = []
+  if (existingAllowMethodsHeader) {
+    methods = existingAllowMethodsHeader.replace(/\s/g, '').split(',').map((method) => method.toUpperCase())
   }
+  methods = uniq([...methods, 'GET', 'POST', 'OPTIONS', 'DELETE'])
 
-  const mergedMethods = uniq([...existingMethods, 'GET', 'POST', 'OPTIONS', 'DELETE'])
+  // if endpoint urls are specified, then we only allow those endpoints
+  // otherwise, we allow any client url to access companion.
```

Comment:
```suggestion
  // If endpoint urls are specified, then we only allow those endpoints.
  // Otherwise, we allow any client url to access companion.
```

maybe like this?